### DISCUSSION TITLE
FT813_Riverdi Display 7 inch

### DISCUSSION THREAD

**Discussion by Overkooked-fff** [RudolphRiedel/FT800-FT813#23] [General]
Hello RudolphRiedel,

I recently bought this LCD : https://riverdi.com/product/rvt70aqffwc00/

And Riverdi click adapter : https://www.mikroe.com/riverdi-click

MCU: NXPS32K148

I want to get works the display with NXPS32K148 mcu . But it does not work yet. 

I would like to use your library but,  I have some questions before all. Is my HW  ok for getting work this display? Which one of your library could be more adapted to my mcu target? 

Thank you in advance for your help.
Best Regard,
Milad

**Comment by RudolphRrr** [comment]
Hello,

the RVT70 is supported, in fact I have one running here one desk over for a week now as HMI for a control unit under test.
But I am not sure regarding the riverdi click.
This is just a passive adapter in a specialised form factor.
If this works would depend on how fast you are driving the SPI and if your controller is capable of providing enough power to drive the SPI at the desired frequency.
And then there might be an issue with providing enough current for both the 3.3V logic-supply and the 5V supply for the backlight.

My adapter boards ( https://github.com/RudolphRiedel/EVE_display-adapter ) are all active and have at least a 3.3V step-down regulator.

I do not have hands on experience with the S32K series (yet) but I do have an eval board with a S32K144.
Right now I do not even know what I would need to build software for it, apart from GCC, obviously, I have not tried it, yet.

What needs to be changed is in EVE_target.h.
I hope the different targets in there provide a good example of what needs to be done, mostly these are hardware specific one-line functions.
In a second step EVE_target.c would be the file in which DMA support is added.

**Reply by Overkooked-fff** [reply]
Hello,

Thank you for your explanation. 

I will try a software for my S32K148 MCU and come back to you. 

First of all, I have another question, what is the role of the PD pin? Is it necessary to operate the display? because I don't have a PD pin on the riverdi click adapter.

BR,
Milad

**Reply by RudolphRrr** [reply]
PD is used by EVE_init() in EVE_commands.c to send the display thru a power-cycle in order to make sure that all registers are resetted to their default values.

At the beginning of EVE_init() there is one line commented out:
/*	EVE_cmdWrite(EVE_CORERST,0); */ /* reset, only required for warm-start if PowerDown line is not used */

**Comment by RudolphRrr** [comment]
I just found the thread on the FTDI community forum and their assesment that the adapter does not match the display is wrong.
I am not sure what pin 9 "MODE" is supposed to be but this is not connected on the display side.
The Riverdi Click is missing pin 10 for Audio out.
Other than that the adapter is a match for the display.

And pin 8 on the Riverdi Click is labeled "RST", this pin is the PD pin on the display, this connects to pin 12 on the EVE chip:
"PN_N   Chip power down mode control input, active low. Connect to MCU GPIO for power management or hardware reset function,"

**Reply by Overkooked-fff** [reply]
Thank you for your update,

- I checked with Riverdi about Click board, they confirmed me that both display and Riverdi click use same Ribus interface so they do match. 

- They said that I should have 5v for backlight and 3.3V for VCC. Shall I connect both power supply to display? or one is enough?

- Another question is that  , Shall I use a GPIO pin for CS# ?or shall I use the CS# of SPI ?

**Comment by RudolphRrr** [comment]
You need both the logic supply and the backlight supply.
Be aware that the display draws up to 130mA from the 3.3V supply and 550mA (at 5V) from the backlight supply, depending of course on the level the backlight is set to and the ratio of active to not-active pixels.

A smaller display like the EVE3-50 from Matrix Orbital could be used with only 3.3V.
And on the other side I am using a RVT101HVBNWC00-B with two step-down regulators and have the one for backlight set to 8.2V since the RVT101 supports a hight backlight voltage and a higher voltage brings down the current.

I am using a GPIO for CS in my library as I am supporting a wide range of controllers of which at least half do not have hardware support for the SPI CS line.
And then there are controllers that do have hardware support for SPI CS line but only for single byte transfers.

**Comment by RudolphRiedel** [comment]
I am trying to make use of the S32K144EVB I have but so far this only provided the worst experience I have ever had with a new eval board.
I had to use a NXP account to download this "free" software with "unlimited" use and had to use an activation code that
I received by mail in order to install it. And this "unlimited" and free license expires in four years.
The "Get Started" page for the S32K144EVB is way outdated.
The S32 Design Studio for ARM V2.2 from 2020 installs a FreeMaster 2.0 software from 2015 which can not connect to my board.
The FreeMaster 3.1.2.4 which I downloaded additionally with an extra activation code can not connect to my board either.
Downloading software from PEmicro also requires a login. And the supplied driver states in the debug settings "Please register your software to remove this message."
The PEmicro page offers version 12.4 of the driver from 2017 while the driver supplied with S32 Design Studio is 12.7 from 2020.
Unbelieveable.
Ok, I am installing S32DS 3.4 now with not much hope for this one either.

**Reply by Overkooked-fff** [reply]
It's good that you have started using the S32K144EVB. I'm sure you will find a solution.
It's weird this difficulty to install SWs, I never had any difficulty to install SWs. 
First you should create an account on NXP website, then you try to install the SW like this:
Design--> All swftware --> S32 design studio for Arm --> Download --> select the latest version and click download. You will then get the activation key to install the software, too. S32 design studio and PEmicro will install on your pc normally.
If you need help to install the software, don't hesitate. 

On my side I tried to integrate your library in S32 design studio but I have an error with EVE_commands.c. It could not be saved, I don't know why...

**Reply by RudolphRiedel** [reply]
Well, I already had an account with NXP and S32DS 3.4 kind of works now.
Kind of, since I can compile an example and put in on the board with "Debug" but just flashing it to the target with "Run" does not work.
I really do not like eclipse and S32DS does nothing to change my mind.

At least I can try to make it compile now.

**Comment by RudolphRiedel** [comment]
Now that it compiles I started to expand the "hello_world_s32k144" example.
The first thing to do is to add code for the target controller which is done mostly in EVE_target.h and partly in EVE_target.c.

In order to do so you need to either find defines for the target controller that are already set in the development enviroment or add defines.
Now we know that the compiler used for S32K is GCC so the new target goes into the section within
#if defined (__GNUC__) / #endif
The build settings for the "hello_world_s32k144" example show defined symbols in the preprocessor section:
CPU_S32K144HFT0VLLT
And going thru the includes in the dependency tree we find "device_registers.h" which has this line:
#if (defined(CPU_S32K148) || defined(CPU_S32K146) || defined(CPU_S32K144HFT0VLLT) || defined(CPU_S32K144LFT0MLLT) || defined(CPU_S32K142) || defined(CPU_S32K142W) || defined(CPU_S32K144W))

So I added a new target in EVE_target.h like this:

/*----------------------------------------------------------------------------------------------------------------*/
/*----------------------------------------------------------------------------------------------------------------*/

#if defined (CPU_S32K148) || (CPU_S32K144HFT0VLLT)

#include "pins_driver.h"
#include <stdint.h>

#define EVE_CS_PORT PTD
#define EVE_CS 15
#define EVE_PDN_PORT PTD
#define EVE_PDN 16
...

It does not do anything usefull so far but it compiles just fine.
The SPI is not initialised, the SPI functions in EVE_target.h are all empty and the pins selected for CS and PDN are the ones from the build-in LEDs on the eval board.
But it calls TFT_init() in main() and compiles fine so far.

Now I have to check the board documentation to find out which pins could be actually used for CS, PDN and the SPI.

[src.zip](https://github.com/RudolphRiedel/FT800-FT813/files/6917002/src.zip)

**Reply by Overkooked-fff** [reply]
Excellent work and thank you. 

I tried it, but I couldn't configure the target and main as well as you did. 

But it still doesn't compile for me. I have a S32K148EVB but I don't think it matters. I still get an error, the error is that it doesn't know #include "sdk_project_config.h" in main.c. 

Is "sdk_project_config.h" already included in the "hello_world_s32k144" example? Or have you developed "sdk_project_config.h"?

A simple way to add and configure SPI driver: 
You can add the SPI driver to the project by double-clicking on the component in the Processor Expert window, then finding the SPI driver and double-clicking on it to add it to the project. 
Then, to configure the SPI driver, you can double-click on the driver you added in the Processor Expert.

**Reply by RudolphRiedel** [reply]
The "sdk_project_config.h" is in the example code. This is an odd one for me as usually the controller includes are used.
Like avr/io.h, sam.h or stm32f3xx.h.
No idea really why NXP has to do this differently.

So just create a new project from the examples and toss the files in there.

I just picked annother example: "lpi2c_master_s32k148".
And it's the same, #include "sdk_project_config.h" early in main.c.
I copied over all the files except main.c to \workspaceS32DS.3.4\lpi2c_master_s32k148\src\ 
Then I added the two include lines
#include "tft.h" and #include "EVE.h"
to the main.c from the example and the call to TFT_init();
right before the file loop in main() and it compiles with no warning while making the binary around 9k bigger.

And I rather not experiment with features in this IDE yet, I already hate it enough as it is. :-)

Ok, annother experiment, what happens when we create a new project?
Then the main.c starts with #include "S32K144.h" and there is no "pins_driver.h".
Good job NXP on providing clean examples...

So well, go bare metall.

static inline void EVE_cs_set(void)
{
	EVE_CS_PORT->PCOR = 1 << EVE_CS;  // active low
}

static inline void EVE_cs_clear(void)
{
	EVE_CS_PORT->PSOR = 1 << EVE_CS; // active high
}

static inline void EVE_pdn_set(void)
{
	EVE_PDN_PORT->PCOR = 1 << EVE_PDN;  // active low
}

static inline void EVE_pdn_clear(void)
{
	EVE_PDN_PORT->PSOR = 1 << EVE_PDN; // active high
}

This still does not do anything usefull so far but it compiles.
[EVE_Test.zip](https://github.com/RudolphRiedel/FT800-FT813/files/6917957/EVE_Test.zip)

**Reply by Overkooked-fff** [reply]
It does not complie first.
I don't know maybe you use different software. I use "S32 Design Studio for ARM Version 2.2". I think you use "S32 Design Studio for S32 Platform".

Then I picked an example into the my workspace and tried to copied the files into the source without main.c. Then configured main.c. 
So , now is compiling without error. 
Thank you :+1:

**Reply by RudolphRiedel** [reply]
I am using S32 Design Studio 3.4 now and it at least kind of worked.
I still have not found out how to simply flash the board.

The next setback I encountered was that NXP is not providing proper CMSIS headers. :-(
So there is no SysTick_Config() available although this is "just" a Cortex M4.

But right now I can not even flash the application with "Debug" anymore.
"Error with command: gbd --version"

Somehow it broke over night without changing anything, not even updating anything.
Worst experience with a new controller ever.

**Reply by Overkooked-fff** [reply]
S32 Design Studio 3.4 is for the S32 platform and I don't know it.

I use "S32 Design Studio for **ARM**", and I have no problem flashing and debugging my EVB.

**Reply by RudolphRiedel** [reply]
How do you just flash with S32DS?
And S32DS 3.4 is for S32K as well, supporting all families.
I can still build for S32K144.
But even after de-installing and re-installing it, "Debug" does not work anymore so I can not even flash example code anymore.

**Reply by Overkooked-fff** [reply]
Sorry for my previous comment, I have no problem **debugging** my EVB :-( 

I'm a new NXP MCU user, too. I don't know how to flash it yet. One of my colleagues knows but unfortunately he's on vacation now.

**Reply by Overkooked-fff** [reply]
I Asked a question about flashing with S32DS in NXP community. I hope that NXP expert could help us.

https://community.nxp.com/t5/S32K/How-to-flash-S32K148-in-S32-design-Studio-for-ARM-2-2/m-p/1317336#M11515

**Reply by RudolphRiedel** [reply]
I am also on vacation now. :-)
Well, somehow I can upload by "Debug" again and I found AN5413.pdf now with bare metal examples.
At least I have the red LED blinking on the S32K144EVB now, still with a busy-loop delay though.
The port_init() I found earlier was missing a write to PORTx->PCR[pin] in order to set up the MUX for GPIO.

**Reply by Overkooked-fff** [reply]
You are working even on vacation ... you deserve kudos :1st_place_medal:

**Reply by RudolphRiedel** [reply]
Except that this is not working, it's not like my employer is paying for any of this anyways. :-)

Systick is up and running now, I just leave the S32K144 at running with the default 48MHz, at least for now.
EVE_init_spi() should be initialising the CS and PDN pins now, I went with PTB5 and PTD14 as these are next to the SPI on the "Arduino" headers.
Nothing on the logic analyzer yet though. 

Please remove these documents again, I am not sure if any of these are freely available and I do have an account with NXP anyways.

**Reply by RudolphRiedel** [reply]
Ok, after hooking up the logic analyzer to the correct pin-header I can see at least CS and PDN work now:
![grafik](https://user-images.githubusercontent.com/31180093/128025872-c3f51883-bff7-4fc5-afa2-4dc3354d0530.png)

**Reply by Overkooked-fff** [reply]
That's nice.

About flashing I could find this : 
To flash the generated code, then inside the folder you will find a .elf file. In S32DS for ARM you have to go to the Flash window, Select a project for S32K144, then Browse for the .elf file and press Flash.

![mariuslucianand_1-1626346772092](https://user-images.githubusercontent.com/70748328/128028185-11a9bda5-964c-42c8-b4fe-1f4732924c8a.png)

**Reply by RudolphRiedel** [reply]
Nope, I found the dialogue in S32DS 3.4 but the flash-symbol is not showing.
Well, "Debug" and "Disconnect" works now.
Hmm, or at least it did a minute ago and now does not anymore, some issue with compiling the code, S32DS is weird.
Shut down S32DS, restarted it and it compiles again...

**Reply by RudolphRiedel** [reply]
Ok, time to take a break.
[EVE_Test.zip](https://github.com/RudolphRiedel/FT800-FT813/files/6925422/EVE_Test.zip)

There is an EVE_init_spi() now in the included EVE_target.c and spi_transmit() in EVE_target.h has code.
But I am not getting anything out of the SPI so far.

**Reply by Overkooked-fff** [reply]
There are some examples for SPI in S32DS . in the photo attached you can find SPI driver and the API for SPI communication. 
![SPI](https://user-images.githubusercontent.com/70748328/128142986-773007be-7dd5-4e97-913f-0f608400ae94.JPG)


Once you click on the SPI driver, you can configure the SPI. And by double clicking on the pin mux, you can configure the pins for GPIO, SPI, ... as well.

**Reply by RudolphRiedel** [reply]
Well, I tried to figure out how the SPI is working in S32K on register level.
And I was looking at example code.

What I find strange in the peripheral driver is that you have to configure it to use IRQ or DMA, there is no way to use neither.
And I removed the slave from the configuration and still ended up with lpspi_slaver_driver.c in the project.
Also, there is no option to not use any of the chip select pins?

Can you get out anything of the SPI using the peripheral driver?

**Reply by Overkooked-fff** [reply]
Well, you can do not use CS pin in SPI , I already have an example and just using SCK and MOSI in SPI. 

Actually, I always use the example of the S32 to do my projects and simply modify it based on my aplication. I don't go into the details of configurations and registers, I use the bas of the example because I'm not so familiar with NXP. So, sorry I do not have any idea about your seconde question. 

Any way, for initialization of clock, SPI and the other driver I use the API already existe in the drivers. But in the example that you shared I found that you initialized SPI and clock in the other way. In the SPI transmit fonction maybe you need to use API of SPI who is found in the SPI driver : 
![SPI2](https://user-images.githubusercontent.com/70748328/128155828-8cdd4828-2b9a-4e08-bd9d-00be959f292d.JPG)

and may be you to initialize clock and SPI like this : 

![SPI3](https://user-images.githubusercontent.com/70748328/128156070-d4fc4e98-96e4-48aa-8b77-fa31885d97c9.JPG)

I'm trying to add your files in a SPI example and try to see if I get the output of SPI .

**Reply by Overkooked-fff** [reply]
once I run the programe I can see some datas on the MOSI by oscilloscope , but I get this error imediately, I don't know why:

![spi4](https://user-images.githubusercontent.com/70748328/128191786-b2f72ae5-fe2a-45cc-a027-db5ff1e51bbd.JPG)

[Sources.zip](https://github.com/RudolphRiedel/FT800-FT813/files/6931720/Sources.zip)

**Reply by Overkooked-fff** [reply]
I increased the size of rxBuff (that I created ) in the spi_transmit : uint8_t rxBuff[256]; 
So the error disapeared ...

I have some data on the SPI but it stops after a few seconds. Maybe it stops after initialization. 
What function/command can I use to display something on the screen?

**Reply by RudolphRiedel** [reply]
Ok, one step back, using the supplied peripheral libraries it is.
I started with the lpspi_transfer_s32k144 example which is supposed to read in the ADC and push out the result on the SPI.
Well, there is traffic on the SPI but all I see is Zero bytes and additionally it configures a second SPI as slave and a PWM output.
I stripped that down to only the master SPI and just increased "masterDataSend" every time.
The SPI configuration does not match the S32K144EVB with SPI_SIN beeing routed to the wrong pin and using the wrong chip-select.
Then I tried to switch over to not using hardware chip-select.

There is no option to not use any of the four chip selects which means that whatever chip select you choose will be controlled by hardware if you are not very carefull.
If you were to use two SPIs you better not use the same chip selects.
And what I also do not get is how this is supposed to work with more than one chip select in use on a single SPI.

Also, there is no option to not use interrupts or DMA which I find rather strange as well.

So at first I changed the chip select used to LPSPI_PCS1 since this one is routed to the pin-header next to the SPI pins.
Then I modified the .mux field in pin_mux.c for this pin to PORT_MUX_AS_GPIO and it stopped working altogether.
Checking with the hello_world example I found that there are fields missing from the pin configuration structure.
So I added these and now it started to work.

![grafik](https://user-images.githubusercontent.com/31180093/128319091-ad25cf85-c924-4245-9066-8337b6a28a65.png)

This is painfully slow though, it takes a whopping 14µs after the manual chip-select before the transfer starts and annother 17µs after the transfer is done. And consequently there is a 31µs gap between transfers.
Is there a coinminer.c in there? :-)

This is exactly why I tend to not use supplied peripheral functions as these are usually wasting a lot of processing time.
But at least it transfers bytes over the SPI, this should provide hints on what I did wrong in my version.

**Reply by RudolphRiedel** [reply]
It is starting to look useable.
![grafik](https://user-images.githubusercontent.com/31180093/128338406-1464e8ac-e00c-4921-9e66-7dd78c61259e.png)

[SPI_s32k144.zip](https://github.com/RudolphRiedel/FT800-FT813/files/6938012/SPI_s32k144.zip)

**Reply by Overkooked-fff** [reply]
Yes, SPI works. 

But nothing on the display yet :-(
 I'm not sure if it wil works with riderdi click board or not :-( 
 I do not have PD on the riverdi click board :-(

**Reply by RudolphRiedel** [reply]
This is just an exercise in getting the SPI to work.
And I just found out that the RTOS that comes with the SPI driver is required, damn. :-(
I am beginning to really dislike NXPs examples and SDK.
I also have seen the answer to your request on the NXP forum, so there is no way to just flash the binary to a controller and it does not even occur to them that this is odd.

**Reply by RudolphRiedel** [reply]
[SPI_s32k144.zip](https://github.com/RudolphRiedel/FT800-FT813/files/6938697/SPI_s32k144.zip)

This one is sending out the init-sequence and the first frame.
At least this looks good in the logic analyzer.
The clock, pin and spi init is solely done thru SDK functions, the EVE_init_spi() is still in there and left untouched for now.
So pin_mux.c, pin_mux.h and peripherals_lpspi_1.c need to be adapted for the S32K148EVB since unfortunately the header J2 has a different pinout and to make things worse the pin functions are not mentioned in the "Understanding the Header/Pinout".

This is not sending out frames in a loop since I am not sure yet how to make use of that required "osif_baremetal.c" that is preventing me from using Systick() but that is not really in use.

**Reply by RudolphRiedel** [reply]
I checked the pinout for the S32K148EVB.
S32K144EVB S32K148EVB
J2.11 PTB2    J2.16 PTB2 SCK
J2.9 PTB3      J2.13 PTB3 SIN
J2.7 PTB4      J2.10 PB1   SOUT
J2.5 PTB5/CS1 PTB0/CS0
J2.3 PTD14    PTA30   EVE_PDN

So yeah, they put a different pin for SOUT on the header and a different chip-select pin.

**Reply by Overkooked-fff** [reply]
Yeah , they are different :-( 

Thank you for the pinout selection :+1:

**Reply by RudolphRiedel** [reply]
Hmm, I hooked up the S32K144EVB to one of my adapter boards and to a display.
And it kind of works, I can see the traffic on the SPI from the display and it looks okay.
But somehow a co-processor fault is detected in EVE_busy() although the data on the SPI does not support this and consequently EVE is send into a software-reset.
Reading from the SPI does not seem to be working properly.

**Reply by RudolphRiedel** [reply]
And it is up and running for the first time now. :-)
![grafik](https://user-images.githubusercontent.com/31180093/128608025-8ecb0f90-bbf3-46b8-9859-ffd6ba7656c7.png)

This is so not ready for release yet, more proof of concept.
The last issue was that my SPI functions read the receive FIFO before the transfer was actually finished and therefore there was a shift by one byte.

[SPI_s32k144.zip](https://github.com/RudolphRiedel/FT800-FT813/files/6949451/SPI_s32k144.zip)

The main.c is a joke and I am not sure what to do with the EVE_init_spi() - probably remove it.
And I consider the NXP SPI driver to be completely unuseable for this type of operation as it is way too slow.

The display in the picture is a RVT50UQQBNWC01 which is the closest one to the RVT70AQ I have here.
I also tried an EVE3-50G and it worked as well.

Next is DMA support.
I am not yet sure if this works with DMA yet since I read somewhere that there is a limit of 5xx bytes for DMA and there might be an issue with the 3 + x * 4 bytes needed to be send as in this is always an odd number.
I hope the 5xx vytes were a limitation of the example code I saw.

**Reply by RudolphRiedel** [reply]
This is interesting:
https://source.codeaurora.org/external/s32support/s32k144_cookbook/tree/

These examples from NXP do not use their peripheral drivers.

**Comment by RudolphRiedel** [comment]
This is a cleaned up version: 
[SPI_s32k144.zip](https://github.com/RudolphRiedel/FT800-FT813/files/6949722/SPI_s32k144.zip)

I removed the odd SPI driver from NXP after reworking the EVE_init_spi() with inspiration from example code from NXP.
I left in the pin configuration code from NXP since this one does not seem to have any dependencies.
After removing the SPI driver from NXP I could remove the unused RTOS functions as well which made it possible again to use the ARM systick timer in my main.c although I am still surprised that NXP is not providing proper CMSIS headers.

And I removed the now unnecessary defines for the pins in EVE_target.h.

Two things are a bit alarming, one is the still rather low clock speed of the SPI.
I have set it up to about 2MHz now.
Usually I would go with 8MHz but for one I find the clock settings rather strange and then I get issues like the touch no longer
working at higher clock speeds.
The other thing is that I need this for SPI transfers now:

static inline void spi_transmit(uint8_t data)
{
	EVE_SPI->SR |= LPSPI_SR_RDF_MASK; /* clear Receive Data Flag */
	EVE_SPI->TDR = data; /* transmit data */
	while((EVE_SPI->SR & LPSPI_SR_RDF_MASK) == 0);
	(void) EVE_SPI->RDR; /* dummy read-access to clear Receive Data Flag */
}

So far I have not found a way to ignore the received data for transmit-only operations.
This is a concern since transferring buffers of up to 4k for a display refresh in DMA mode is ideally a write-only operation.

**Reply by RudolphRiedel** [reply]
To answer myself, I found RXMSK in the TCR register now which, when set, disables the SPI receiver.

As a pre-cursor to DMA this version is running like it is supporting DMA but the EVE_start_dma_transfer() function
is actually sending the DMA buffer in write-only blocking mode.

[SPI_s32k144.zip](https://github.com/RudolphRiedel/FT800-FT813/files/6951095/SPI_s32k144.zip)

![grafik](https://user-images.githubusercontent.com/31180093/128638573-97a75957-01c1-4101-9798-0e99b93c3591.png)
![grafik](https://user-images.githubusercontent.com/31180093/128638610-0f2f0a43-a210-4d9e-b8fe-6e0e433c69e8.png)
![grafik](https://user-images.githubusercontent.com/31180093/128638642-c64241ea-1ee1-4142-b7e3-df43d211e63d.png)

The first shot is an overview of one TFT_touch() call followed by one TFT_display() call.
The second one is a normal transfer with activated read and the third picture shows the difference when switching over to write-only.

Does this make a whole lot of a difference?
No, not really, sending the 223 bytes of the demo code takes 1.69786 ms in write/read mode and 1.56232 ms in write only mode.
But I found this was interesting enough to share, the pause between the SPI transfers is not really caused alone by the core of the controller running the loop but the the state-machine in the SPI unit is contributing significantly to it.

I also reduced TCR/PRESCALE now from 2 to 1 and then 0.
And the pause between transfers in write/read mode got shorter.
From 1.6µs to 1.1µs to 0.8µs.
In the write only tranfers the pause also got shorter.
From 540 ns to 260ns to 120 ns.

Running the SPI at ~8MHz now reduces the time needed to transfer the buffer to 391µs.

**Comment by RudolphRiedel** [comment]
Damn, I am stuck with this DMA engine.

As a next step this one should transfer the buffer with DMA and wait for completion.
It kind of works, the transfer is started alright and after completion the program continues.

I am only getting one byte of the DMA though.

[SPI_s32k144.zip](https://github.com/RudolphRiedel/FT800-FT813/files/6957042/SPI_s32k144.zip)

The SPI is setup for byte transfers and so I configured the DMA to use bytes as well.
The DMA engine works with a two loop systems, minor loops and major loops.
The minor loop is setup to a transfer size of "1" now and the amount of major loops
is setup to the size of the buffer in bytes.
But somehow it stops after just one minor loop.

When I increase the amount of bytes per minor loop more bytes get transferred on the SPI.
But only a fraction of the amount setup and the bytes that are send are not what is in the buffer.
I  just setup 122 bytes and the result is 11 bytes on the SPI.
1 -> 1
2 -> 2
3 -> 3
4 -> 4
5 -> 4
6 -> 4
7 -> 5
8 -> 5
9 -> 5
10->5
11->5
12->5
13->5
14->5
15->5
16->5
20->5
22->5
23->5
24->6
30->6

What is this DMA engine doing and why?

**Comment by RudolphRiedel** [comment]
I got some help from NXP or rather they pointed me in the direction of annother SPI/DMA example I was not aware of so far.

And this is the first version that transfers the buffer for the screen update with DMA:
[EVE_Test_S32K144.zip](https://github.com/RudolphRiedel/FT800-FT813/files/6964195/EVE_Test_S32K144.zip)

It still does busy-wait for the DMA to complete, to finish this I still need to add the DMA-done interrupt.

**Comment by RudolphRiedel** [comment]
Now this was easier than I thought it might be.

This one is using an DMA-done interrupt:
[EVE_Test_S32K144.zip](https://github.com/RudolphRiedel/FT800-FT813/files/6964392/EVE_Test_S32K144.zip)

Edit, I merged the changes and released new versions of EVE_target.c and EVE_target.h.

**Reply by Overkooked-fff** [reply]
Thank you for your updates and great job.
I was traveling and couldn't follow you.

**Reply by RudolphRiedel** [reply]
This was an interesting opportunity to actually do something with the S32K144 I bought a couple of years ago and at the same time I know we currently have a project in my department at work with a S32K14x (that I not really was involved in, yet).
If I am asked this is annother family I played with. :-)
